### DISCUSSION TITLE
Form submission with double opt-in (DOI) e-mail verification

### DISCUSSION THREAD

**Discussion by wnzirngibl** [statamic/cms#8828] [Q&A]
Hello community,
we're trying to implement a kind of subscription workflow for anonymous users with Statamic forms.

One necessary feature is the validation of the user's e-mail address by a double opt-in verification mail - so the form must not be submitted before the e-mail address is validated.

So the desired workflow would be:

1. user fills out Statamic form
2. user gets verification mail to the e-mail address they entered
3. user clicks on verification link in e-mail
4. the form is submitted

Being rather new to Statamic I was hoping to find something like that in core or an add-on, but I couldn't find very much.

I found something in the Laravel docs about their user verification process: https://laravel.com/docs/10.x/verification

At the moment, that would be the best guess, trying to implement something along these lines - although we don't really need the form data as users, it's really only about verifying the e-mail address before sending out the form.

If anybody has ideas about how to achieve this with Statamic, help would be greatly appreciated.

**Comment by ryanmitchell** [comment]
You could hook into the events Statamic provides for submissions (eg SubmissionCreated), and mark the Submission as not validated. Then use the email functionality already there to give a link to a custom route for verification that then marks it as validated.

**Reply by wnzirngibl** [reply]
Thanks a lot for your ideas! Sounds like a promising way to go.

However, I don't have any idea how to provide this custom route for verification - is that something generated by Statamic, or would I have to implement it by myself? Is there any documentation about the email functionality you are mentioning?

**Reply by ryanmitchell** [reply]
https://statamic.dev/forms#emails

You would have to implement the route yourself

**Reply by wnzirngibl** [reply]
OK, thanks. I'll look into it.

**Comment by mariohbrino** [comment]
I have implemented a solution to handle the verification during the form submission, I've thought to create a Statamic Addon.
